src/app/rundown-view/services/action-trigger-producer-key-binding.service.ts

```diff
@@ -89,14 +91,35 @@ export class ActionTriggerProducerKeyBindingService implements KeyBindingService
     this.actionTriggersWithAction.clear()
 
     this.actionTriggers.forEach(actionTrigger => {
-      const actionForTrigger: Tv2Action | undefined = this.actions.find(action => action.id === actionTrigger.actionId)
-      if (!actionForTrigger) {
+      let action: Tv2Action | undefined = this.actions.find(action => action.id === actionTrigger.actionId)
+      if (!action) {
         return
       }
-      this.actionTriggersWithAction.set(actionTrigger, actionForTrigger)
+      if (action.type === PlaceholderActionType.CONTENT) {
+        action = this.findActionForContentPlaceholderAction(action as Tv2ContentPlaceholderAction, actionTrigger)
+      }
+      if (!action) {
+        return
+      }
+      this.actionTriggersWithAction.set(actionTrigger, action)
     })
```

Comment:
Instead of having the logic for determining when to resolve the action and the logic for updating in the same method in the same method, they could be separated, which would lead to an immutable flow for making the decisions. E.g.
```typescript
private getResolvedAction(actionTrigger: ActionTrigger<KeyboardTriggerData>): Tv2Action | undefined {
  const action: Tv2Action | undefined = this.actions.find(action => action.id === actionTrigger.actionId)
  if (action?.type !== PlaceholderActionType.CONTENT) {
    return
  }
  return this.findActionForContentPlaceholderAction(action as Tv2ContentPlaceholderAction, actionTrigger)
}
```

Suggestion: `actionTriggersWithAction` is updated via side-effects in the forEach. An alternative solution could be  to use reduce to generate the entries for a new Map.
```typescript
this.actionTriggersWithAction = New Map(
  this.actionTriggers.reduce((entries, actionTrigger) => {
    const action: Tv2Action | undefined = this.getActionForActionTrigger(actionTrigger)
    return action ? [...entries, [actionTrigger, action]] : entries
  }, [])
)
```

Reply:
I'm not sure I follow your first sentence. The logic for determing when to resolve is not in this method. When calling `mapActionTriggersToActions` it has _already_ been decided that now we should "resolve" them and this method has nothing to do with that logic.

For your `.reduce` suggestion: did you try it? I couldn't get it to compile.

Reply:
The code from line 92 to 103, determines the action and whether it should be resolved based on the action type. The change from `const actionForTrigger` to `let actionForTrigger` is an indicator that some conditional logic is added to the method. By moving the computation to a separate method, the decision making that the introduced logic is responsible for is now separated from this method and we can keep it as an immutable computation.

Sorry, the reduce was not tested – my bad. I have added code that compiles below. It is just a suggestion for an immutable data change flow.
```typescript
  private actionTriggersWithAction: Map<ActionTrigger<KeyboardTriggerData>, Tv2Action> = new Map()

   // ...

    this.actionTriggersWithAction = new Map(
      this.actionTriggers.reduce((entries, actionTrigger) => {
        const action: Tv2Action | undefined = this.getResolvedAction(actionTrigger)
        return action ? [...entries, [actionTrigger, action]] : entries
      }, [] as [ActionTrigger<KeyboardTriggerData>, Tv2Action][])
    )
```

Reply:
I have extracted the logic for finding the Action to it's own separate method.